staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -389,10 +390,7 @@ func (f *DeltaFIFO) queueActionLocked(actionType DeltaType, obj interface{}) err
 	} else {
 		// This never happens, because dedupDeltas never returns an empty list
 		// when given a non-empty list (as it is here).
-		// But if somehow it ever does return an empty list, then
-		// We need to remove this from our map (extra items in the queue are
-		// ignored if they are not in the map).
-		delete(f.items, id)
+		panic(fmt.Sprintf("Impossible dedupDeltas for id=%q: old items=%#+v, obj=%#+v", id, f.items[id], obj))
```

Comment:
Let's not panic - if because of some reason we have a bug and this would happen, we would never recover from crashloop in such case.
Instead let's simply return an error here.

Reply:
I have had others advise me that the panic is actually the preferred response.  I think the theory goes something like the following.  The question is how can code react when it detects data corruption that must be due to a bug that presumably bites in rare circumstances (if it bit commonly then the code would not have made it past CI, and there is no response that will save the day).  If the in-memory data is now corrupted then keeping the process alive with the corrupted data is going to run into more trouble than restarting the process and letting it run successfully until such time, if any, when the bug bites again.

Reply:
I don't think I fully agree.

In cases like this, data-corruption basically happens due to some bug in the code. We are all just humans - while we believe it should never happen, noone knows if there isn't any subtle bug still around.
So if it happened once, it can also be repeatable and we can enter the crashloop.

So while in test environment it would be desired to panic (and even enter crashloop), it production, I would like to get the situation where because of bug in the code I'm not able to exit this crashloop (with the current state of etcd) and the only way to fix my production cluster is to fix the bug and rebuild it.

So I'm really opposed to switching to panics here.

@liggitt @lavalamp - for thoughts

Reply:
Failing fast vs tolerance against errors is always controversial.  :)

I guess the decision is case by case. It depends on how critical maintaining such data invariant is such that we are willing to shut down the entire system if we end up breaking that invariant.

Reply:
It seems to me that the only good answer is to write code to repair the data structure.  In this case I think that would be fairly straightforward, no missing critical information to guess at.

Reply:
Latest revision repairs instead of panicing or returning an error.